sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -388,11 +388,15 @@ cdef class Splitter:
                     &left_indices_buffer[offset_in_buffers[thread_idx]],
                     sizeof(unsigned int) * left_counts[thread_idx]
                 )
-                memcpy(
-                    &sample_indices[right_offset[thread_idx]],
-                    &right_indices_buffer[offset_in_buffers[thread_idx]],
-                    sizeof(unsigned int) * right_counts[thread_idx]
-                )
+                if right_counts[thread_idx] > 0:
+                    # if n_threads >= 2 one has right_counts[-1] = 0 and
+                    # right_offset[-1] = len(sample_indices) leading to
```

Comment:
I'm likely missing something but I don't understand these 2 lines

in the run-down example above, we have n_threads >= 2 and yet right_counts[-1] != 0 and right_offset[-1] != len(sample_indices)

Also it's not clear to me why this issue only ever happens for the right node?

Reply:
Didn't you write that code? 😏 

The "last" thread is responsible for the most "right" samples. The "right" part of the "last" thread sorts the last indices. In can happen that `right_counts[-1]=0`. Then the `memcpy` is a no-op. But it also leads to potentially evaluating the first argument of `memcpy`, which is `&sample_indices[right_offset[thread_idx]] = &sample_indices[n_samples]` which accesses one element after the last one, i.e. out of bound.

Again, this is a no-op, but it should not be executed for potentially undefined behavior.

Reply:
I updated the comment in the hope to make it clearer.